packages/client-sdk-web/src/internal/pubsub-client.ts

```diff
@@ -0,0 +1,478 @@
+import * as pubsub from '@gomomento/generated-types-webtext/dist/CachepubsubServiceClientPb';
+import * as cachepubsub_pb from '@gomomento/generated-types-webtext/dist/cachepubsub_pb';
+import {Configuration} from '../config/configuration';
+import {
+  CredentialProvider,
+  MomentoErrorCode,
+  MomentoLogger,
+  SubscribeCallOptions,
+  TopicItem,
+  UnknownError,
+} from '@gomomento/sdk-core';
+import {
+  Request,
+  StreamInterceptor,
+  UnaryInterceptor,
+  UnaryResponse,
+  RpcError,
+  StatusCode,
+} from 'grpc-web';
+import {TopicClientProps} from '../topic-client-props';
+import {version} from '../../package.json';
+import {Header, HeaderInterceptorProvider} from './grpc/headers-interceptor';
+import {
+  truncateString,
+  validateCacheName,
+  validateTopicName,
+} from '@gomomento/sdk-core/dist/src/internal/utils';
+import {normalizeSdkError} from '@gomomento/sdk-core/dist/src/errors';
+import {TopicPublish, TopicSubscribe} from '../index';
+import {cacheServiceErrorMapper} from '../errors/cache-service-error-mapper';
+import {SubscriptionState} from '@gomomento/sdk-core/dist/src/internal/subscription-state';
+import {IPubsubClient} from '@gomomento/sdk-core/dist/src/internal/clients';
+
+/**
+ * Encapsulates parameters for the `sendSubscribe` method.
+ */
+interface SendSubscribeOptions {
+  cacheName: string;
+  topicName: string;
+  onItem: (item: TopicItem) => void;
+  onError: (
+    error: TopicSubscribe.Error,
+    subscription: TopicSubscribe.Subscription
+  ) => void;
+  subscriptionState: SubscriptionState;
+  subscription: TopicSubscribe.Subscription;
+}
+
+/**
+ * Encapsulates parameters for the subscribe callback prepare methods.
+ */
+interface PrepareSubscribeCallbackOptions extends SendSubscribeOptions {
+  /**
+   * The promise resolve function.
+   */
+  resolve: (
+    value: TopicSubscribe.Response | PromiseLike<TopicSubscribe.Subscription>
+  ) => void;
+  /**
+   * Whether the stream was restarted due to an error. If so, we skip the end stream handler
+   * logic as the error handler will have restarted the stream.
+   */
+  restartedDueToError: boolean;
+  /**
+   * If the first message is an error, we return an error immediately and do not subscribe.
+   */
+  firstMessage: boolean;
+}
+
+export class PubsubClient<
+  REQ extends Request<REQ, RESP>,
+  RESP extends UnaryResponse<REQ, RESP>
+> implements IPubsubClient
+{
+  private readonly client: pubsub.PubsubClient;
+  private readonly configuration: Configuration;
+  private readonly credentialProvider: CredentialProvider;
+  // private readonly unaryRequestTimeoutMs: number;
+  // private static readonly DEFAULT_REQUEST_TIMEOUT_MS: number = 5 * 1000;
+  private readonly logger: MomentoLogger;
+  private readonly authHeaders: {authorization: string};
+  private readonly unaryInterceptors: UnaryInterceptor<REQ, RESP>[];
+  private readonly streamingInterceptors: StreamInterceptor<REQ, RESP>[];
```

Comment:
Are these a strictly gRPC-web feature? A zoom call could help me understand this part.